[WIP] Webapp feature for public data export

whoodes opened the pull request:
## What is this PR for
Fixes: #11930, which adds a public only realm export feature for the webapp.

*Demo of tentative UI*:
![demo](https://user-images.githubusercontent.com/39782863/62760765-7370c180-ba20-11e9-8ab7-b59330b538c0.gif)



## Commits

### Previous commits
* *(merged)*: This commit serves as the first step in supporting "public export" as a
webapp feature.  The refactoring was done as a means to allow calling
the export logic from elsewhere in the codebase.

* *(merged)*: This allows for removal of the local tarball upon a successful s3 export.

* *(merged)*: An endpoint was created in zerver/views.  Basic rate-limiting was
implemented using RealmAuditLog. The event is then published to the ```deferred_work``` queue processor to prevent the export process from being killed after 60s. Finally, upon completion of the export the realm admin(s) are notified via ```send_event```.

* (Plus a handful of clean up commits).

**Complete** *(re-posting of Tim's comment below)*:
- [x] We need to add LOCAL_UPLOADS_DIR support to the --upload-to-s3 option. First, a commit should rename it to --upload since it's no longer going to be just S3. And then a commit that changes export_realm_wrapper to support upload via the LOCAL_UPLOADS_DIR backend (likely, this will be just a shutil.copyfile call with the right construction), and then we can remove the AssertionError and fill out the test for the LOCAL_UPLOADS_DIR backend.

- [x] There are several TODOs in test_realm_export.py that need to be completed.

- [x] We probably should arrange to store the ID of the RealmAuditLog row created in public_only_realm_export in the event sent to the deferred_work queue, and then have the queue processor add the public_url to the RealmAuditLog event in the extra_data field. This will enable support for deleting the uploaded exports later once they've been accessed.

hackerkid commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,52 @@
+    # A basic implementation for rate-limiting realm exports
+    # We check that the number of events taking place within our delta does not exceed the limit
+    realm_exports = RealmAuditLog.objects.filter(realm=realm, event_type=event_type)
+    realm_exports = [log for log in realm_exports if event_time > event_time_delta]
```
Shouldn't this by `log.event_time`?

hackerkid replied:
Shouldn't this by log.event_time?
Also this won't be required if we do it the way I described.

hackerkid commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,52 @@
+
+    # A basic implementation for rate-limiting realm exports
+    # We check that the number of events taking place within our delta does not exceed the limit
+    realm_exports = RealmAuditLog.objects.filter(realm=realm, event_type=event_type)
```
You can directly get the last 5 events by doing something like this

```
last_5_logs = RealmAuditLog.objects.filter(realm=realm, event_type=event_type).order_by("-event_time")[:5]
```
and see if last_5_logs[4] is greater than event_time_delta or not.

This can probably be even simplified. Also double check the query I wrote is correct.

hackerkid replied:
You can directly get the last 5 events by doing something like this

```
last_5_logs = RealmAuditLog.objects.filter(realm=realm, event_type=event_type).order_by("-event_time")[:5]
```
and see if last_5_logs[4].event_time is greater than event_time_delta or not.

This can probably be even simplified. Also double check the query I wrote is correct. I wrote the pseudo code to convey the idea.

hackerkid commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,52 @@
+        return json_error(_('Exceeding weekly organization export limit'))
+
+    output_dir = tempfile.mkdtemp(prefix="zulip-export-")
+    if os.path.exists(output_dir):
```
Don't check whether the path exist like this. It can lead to race conditions. Just delete the directory directly and ignore the exception that arise in case a directory exist using `pass`.

hackerkid commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,52 @@
+             'time': event_time,
+             'realm_name': realm.name}
+    queue_json_publish("deferred_work", event)
+    command_independent_realm_export(realm=realm, output_dir=output_dir,
```
So basically we add things that we dont want to instantly execute to a queue. We add the item by calling `queue_json_publish("deferred_work", event)`. The queue worker will take the items in the queue one at a time and execute the appropriate function. The worker for `deferred_work` queue is defined in `DeferredWorker` in `queue_processors.py` file. You should call `command_independent_realm_export` function there and not here. If you call the function here it would be executed immdeatly. We dont want that. Also move the directory creation code to queue worker as well.,

hackerkid commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,52 @@
+                                     delete_after_upload=True)
+
+    export_log = []
+    log = RealmAuditLog(realm=realm,
```
This can be just
```
RealmAuditLog.objects.create(realm=realm,
                        event_type=event_type,
                        event_time=event_time)
```

hackerkid commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,52 @@
+    RealmAuditLog.objects.bulk_create(export_log)
+
+    # Notify realm admin(s) of a successful export
+    admin_ids = [user.id for user in realm.get_admin_users()]
```
This should go in the queue worker as well.

hackerkid commented:
@whoodes I have commented how to use queue system. Let me know in CZO if there is something that you did't understand. Also posted a bunch of minor comments.

hackerkid commented on zerver/lib/export_realm.py:
```diff
@@ -0,0 +1,51 @@
+from zerver.lib.utils import generate_random_token
+from zerver.models import Realm
+
+def command_independent_realm_export(realm: Realm, output_dir: str, threads: int,
```
I think we should come up with some better name for this function.

whoodes commented:
Ready for another round whenever you are, @hackerkid .

timabbott commented:
@whoodes thanks for working on this!  I have a few proposed changes:
* Let's move the function you put in `zerver/lib/export_realm.py` in the existing `zerver/lib/export.py`.  And I think the name isn't ideal; how about `export_realm_wrapper` ?  It's a bit tricky to name this given the existing `do_export_realm`, but I think that's clear enough.
* With that change, I think the first two commits are safe to merge.  
* For the third commit, add `@require_realm_admin` on the view function, and test that it's only usable by organization administrators.
* I'll have more suggestions, but this is a good start.

timabbott commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,35 @@
+    # We check if the nth from the last record's event time occurred within the time delta
+    # If true, the request to export the realm exceeds the weekly limit
+    limit_check = RealmAuditLog.objects.filter(realm=realm,
+                                               event_type=event_type).order_by("-event_time")[:5].last()
```
Checking the last 5 RealmAuditLog events doesn't make sense -- it'll create potential security issues.  We should just filter the database query with e.g. `event_time__gte=event_time_delta`).

timabbott commented on zerver/models.py:
```diff
@@ -2399,6 +2399,7 @@ class RealmAuditLog(models.Model):
     REALM_SCRUBBED = 'realm_scrubbed'
     REALM_PLAN_TYPE_CHANGED = 'realm_plan_type_changed'
     REALM_LOGO_CHANGED = 'realm_logo_changed'
+    REALM_EXPORT_TRIGGERED = 'realm_export_triggered'
```
Let's call this `realm_exported` (and `REALM_EXPORTED`, etc.)

timabbott commented on zerver/worker/queue_processors.py:
```diff
@@ -609,3 +613,25 @@ def consume(self, event: Mapping[str, Any]) -> None:
                 (stream, recipient, sub) = access_stream_by_id(user_profile, stream_id,
                                                                require_active=False)
                 do_mark_stream_messages_as_read(user_profile, client, stream)
+        elif event['type'] == 'realm_export_triggered':
```
this should be `realm_exported`.

timabbott commented on zerver/worker/queue_processors.py:
```diff
@@ -609,3 +613,25 @@ def consume(self, event: Mapping[str, Any]) -> None:
+                    elif os.path.isdir(file_path):
+                        shutil.rmtree(file_path)
+                except FileNotFoundError:  # nocoverage
+                    pass
```
I don't understand the purpose of this loop -- if you made a secure temporary directory, it should be empty already.

whoodes replied:
Thank you, looking at it now I don't understand its purpose either.

timabbott commented:
> Testing for the admin notification event.

This should be tested in `test_events.py` similar to our other tests, though it'll be a simple one (since the event is a no-op).

whoodes commented:
Thanks for reviewing @timabbott, I addressed the current batch of comments.  I made some additional updates based upon working on the frontend piece of this as well.  A couple questions:
* Do we want to send a ```json_error``` if the s3 backend isn't configured?
* I changed the admins to be notified via a PM from notification bot, curious as to your thoughts on this approach (or how you would prefer notifying admins).

Looking forward to it.

whoodes commented:
(Just rebased from master.)

whoodes commented:
Added the top level page to keep this PR even with where I am at locally. 

 I am getting some thread issues when hitting this from the frontend.  Not sure if it is because ```export_realm_wrapper``` is throwing a ```CommandError``` from not having an s3 backend configured, and that is causing the queue processor to have problems handling the data on the queue.

timabbott commented:
I merged the first couple commits, since they looked complete.  Can you post the precise threading error traceback you got?  I'll need that to help figure out what's going on.

> Do we want to send a json_error if the s3 backend isn't configured?

`json_error` is for situations which are a user error (HTTP code 4xx), not for system configuration problems (5xx).  I think what we need to do instead is extend the upload logic to support using the "local file uploads" backend as well (where we effectively just copy the file into an appropriate path under `LOCAL_UPLOADS_DIR`).

timabbott commented on zerver/tests/test_realm_export.py:
```diff
@@ -0,0 +1,62 @@
+        self.login(admin_email)
+        result = self.client_post('/json/exports/data_export')
+        self.assert_json_success(result)
+        self.logout()
```
You never need to logout at the end of tests -- the database (etc.) state is reset at the end of every test case.

timabbott commented on zerver/tests/test_realm_export.py:
```diff
@@ -0,0 +1,62 @@
+class RealmExportTest(ZulipTestCase):
+    def test_hit_api_endpoint(self) -> None:
+        admin = self.example_user('iago')
+        admin_email = admin.email
```
Probably not worth a separate variable for this, just access `admin.email` on the next line.

timabbott commented on zerver/tests/test_realm_export.py:
```diff
@@ -0,0 +1,62 @@
+                                         event_time=event_time))
+        RealmAuditLog.objects.bulk_create(exports)
+
+        msg = 'Exceeding weekly organization export limit!'
```
again, not worth an extra variable for this -- just pass the string into `assert_json_error` like our other tests do.

timabbott commented on zerver/tests/test_realm_export.py:
```diff
@@ -0,0 +1,62 @@
+        result = self.client_post('/json/exports/data_export')
+        self.assert_json_success(result)
+        self.logout()
+
```
Finally, this test is incomplete.  We checked the endpoint succeeded, but not that event was queued or the export actually happened (and I think it will have, because the default model for our queue processors in the test suite is to immediately run the `consume` function for the queue processor when receiving an event, so that you can test the whole flow in a single unit test).  

We possibly don't want this to actually run the full export, but I think it makes sense to mock e.g. `do_export_realm` with `mock.patch` and verify the arguments passed into that using `m.call_args_list` (grep for it to see examples) so that we're testing all the code outside the core export logic here.

timabbott commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,40 @@
+    limit_check = RealmAuditLog.objects.filter(event_type=event_type,
+                                               event_time__gte=event_time_delta)
+    if len(limit_check) >= time_delta_limit:
+        return json_error(_('Exceeding weekly organization export limit!'))
```
Let's make this more similar to existing error messages and more likely to be reusable.  I think "Exceeded rate limit." is actually fine for this.

timabbott commented on zproject/urls.py:
```diff
@@ -381,7 +382,11 @@
+        {'GET': 'zerver.views.video_calls.get_zoom_url'}),
+
+    # Used for public-only realm exporting
+    url(r'^exports/data_export$', rest_dispatch,
```
Let's make the URL `POST /export/realm`

timabbott commented:
Posted another batch of comments on the backend, thanks for your continued work on this!  @shubhamdhama or @pragatiagrawal31 may be able to do a quick review of the frontend part.

whoodes commented:
Hi @timabbott this is ready for another round of review :)

whoodes commented on zerver/tests/test_events.py:
```diff
@@ -2607,6 +2608,23 @@ def do_upload() -> None:
+
+        events = self.do_test(
+            lambda: do_send_notify_admins_event(self.user_profile.realm, False, [self.user_profile]),
+            state_change_expected=False)
```
I'm not entirely sure if `False` here is correct.

timabbott replied:
docs/subsystems/events-system.mdyes, `state_change_expected` should be false here.  You can read more about this in `docs/subsystems/events-system.md` on ReadTheDocs.

whoodes commented on zerver/tests/test_realm_export.py:
```diff
@@ -0,0 +1,77 @@
+        with patch('zerver.lib.export.do_export_realm'):
+            with override_settings(LOCAL_UPLOADS_DIR='/var/uploads'):
+                result = self.client_post('/json/export/realm')
+        self.assert_json_success(result)
```
I'm wondering if a simple success assertion is enough here, or if I should check the `args_list` as well.

timabbott replied:
Yes, we should check `args_list` here as well.  As a general rule, when you're mocking a function, you generally want some assertion about the function being called correctly, and this is the right choice for verifying this code.

shubhamdhama commented:
@pragatiagrawal31 ahh sorry I forgot to review this one, can you review this?

pragatiagrawal31 commented on static/js/settings_org.js:
```diff
@@ -1112,6 +1112,38 @@ exports.build_page = function () {
+            },
+            error: function (response) {
+                var error_message = "Oops! " + response.responseJSON.msg;
+                update_export_status(i18n.t(error_message), true);
```
Just wondering why didn't we used `ui_report.success` and `ui_report.error` directly over here, it would make the code more understandable and synchronized with the other settings. Moreover, I don't think "Oops!" should be the error message over here, it should be something more clear and precise.

whoodes replied:
> Just wondering why didn't we used ui_report.success and ui_report.error

Great to know, thanks! I wasn't aware.

> Moreover, I don't think "Oops!" should be the error message over here

Certainly, I was skeptical of this being acceptable.  Just wanted to wait for insight from a reviewer :) Thanks @pragatiagrawal31

pragatiagrawal31 commented:
I posted a query and a suggestion about the success and error message part, rest the frontend part LGTM!! 👍  
@timabbott FYI.

whoodes commented:
@timabbott I updated the tests.  I added a `setUp()` method to mitigate redundancy a little.  The two tests `test_endpoint_s3` and `test_endpoint_local_uploads` that I have now are practically identical aside from their respective wrappers.  I am wondering if there is a cleaner way to go about this, or if this acceptable given the nature of what is being tested?

timabbott commented on zerver/worker/queue_processors.py:
```diff
@@ -609,3 +611,33 @@ def consume(self, event: Mapping[str, Any]) -> None:
+
+            to_s3 = False
+            if settings.LOCAL_UPLOADS_DIR is None:
+                to_s3 = True
```
We should be passing `upload_to_s3` unconditionally; and then please do a follow-up commit to support uploading with the LOCAL_UPLOADS_DIR upload backend as well.

```
--- a/zerver/worker/queue_processors.py
+++ b/zerver/worker/queue_processors.py
@@ -612,25 +612,21 @@ class DeferredWorker(QueueProcessingWorker):
                                                                require_active=False)
                 do_mark_stream_messages_as_read(user_profile, client, stream)
         elif event['type'] == 'realm_exported':
-            realm = get_realm(event['realm_name'])
+            realm = Realm.objects.get(id=event['realm_id'])
             output_dir = tempfile.mkdtemp(prefix="zulip-export-")
 
-            to_s3 = False
-            if settings.LOCAL_UPLOADS_DIR is None:
-                to_s3 = True
+            # TODO: Add support for the LOCAL_UPLOADS_DIR uploads
+            # backend in export_realm_wrapper so we don't need this assertion.
+            assert settings.LOCAL_UPLOADS_DIR is None
 
-            export_realm_wrapper(realm=realm, output_dir=output_dir,
-                                 upload_to_s3=to_s3, threads=6, public_only=True,
-                                 delete_after_upload=to_s3)
+            public_url = export_realm_wrapper(realm=realm, output_dir=output_dir,
+                                              upload_to_s3=True, threads=6, public_only=True,
+                                              delete_after_upload=True)
+            assert public_url is not None
 
             admins = [user for user in realm.get_admin_users()]
-            content = "A public-only export was successfully completed!\n"
-            if to_s3:
-                content += (
-                    "The data was uploaded to the following url:\n\n"
-                    "http://%s.amazonaws.com/%s/%s" %
-                    (settings.S3_REGION, settings.S3_AUTH_UPLOADS_BUCKET, settings.S3_KEY,)
-                )
+            content = "Your data export is complete and has been uploaded here:\n\n%s" % (
+                public_url,)
 
             for user_profile in admins:
                 internal_send_private_message(
```

timabbott commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,40 @@
+    # Using the deferred_work queue processor to avoid killing the process after 60s
+    event = {'type': event_type,
+             'time': event_time,
+             'realm_name': realm.string_id,
```
We should pass the realm by ID, not name, because that's robust to changes in realm string ID.  This is a general best practice -- it's almost always better to pass objects by numeric, unique ID.

timabbott commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,40 @@
+    queue_json_publish('deferred_work', event)
+
+    RealmAuditLog.objects.create(realm=realm,
+                                 realm_id=realm.id,
```
This `realm_id` field is duplicated.  Read about how Django ForeignKey works -- the `realm_id` field is an implicit field that's a way of setting the value of the `realm` ForeignKey by ID.

timabbott commented on zerver/views/public_export.py:
```diff
@@ -0,0 +1,40 @@
+    # Filter based upon the number of events that have occurred in the delta
+    # If we are at the limit, the incoming request is rejected
+    limit_check = RealmAuditLog.objects.filter(event_type=event_type,
+                                               event_time__gte=event_time_delta)
```
This query should filter on the realm!  Otherwise you're rate limiting export for everyone, whereas we want a per-organization limit.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5594,3 +5594,11 @@ def get_zoom_video_call_url(realm: Realm) -> str:
+                 sender=settings.NOTIFICATION_BOT,
+                 recipient_ids=admin_ids,
+                 to_s3=to_s3)
+    send_event(realm, event, admin_ids)
```
Changed this to just notify the user who did the request.

timabbott commented on static/templates/settings/data-exports-admin.handlebars:
```diff
@@ -0,0 +1,19 @@
+    </div>
+    <form class="form-horizontal">
+        <p class="alert-word-settings-note">{{t "Upload the organization's public-only data directly to  an S3 backend.
+        Or, if no S3 configuration exists, the data will be stored on disk in the specified directory on the Zulip server."}}</p>
```
Change this description to not talk about the S3/LOCAL_UPLOADS_DIR issue -- we're going to make it work regardless.

I think actually this whole alert can be replaced with "This tool does a <a href="/help/export-your-organization">public export</a> of your organization's history."

timabbott commented on static/templates/settings/data-exports-admin.handlebars:
```diff
@@ -0,0 +1,19 @@
+            <div class="wrapper">
+                <div class="settings-section-title">{{t "Export the organization's public-only data" }}</div>
+                <button type="submit" class="button rounded sea-green" id="export-data">
+                    {{t 'Export data' }}
```
Let's call this button "Export organization"

timabbott commented on zerver/tests/test_realm_export.py:
```diff
@@ -0,0 +1,104 @@
+
+    @use_s3_backend
+    def test_endpoint_s3(self) -> None:
+        admin = self.setUp()
```
This function is called automatically even if you don't call it; read about `setUp` methods in the Python unit test system.  I redid this to not have this weird structure, though further cleanup would be useful.

timabbott commented on zerver/worker/queue_processors.py:
```diff
@@ -609,3 +611,33 @@ def consume(self, event: Mapping[str, Any]) -> None:
+                content += (
+                    "The data was uploaded to the following url:\n\n"
+                    "http://%s.amazonaws.com/%s/%s" %
+                    (settings.S3_REGION, settings.S3_AUTH_UPLOADS_BUCKET, settings.S3_KEY,)
```
This line is broken, bad style, and a serious security issue!  As I mention below, it's bad style because it duplicate a computation already happening in `export_realm_wrapper`.  It's broken because the URL you construct here doesn't match what `export_realm_wrapper` does, and a serious security issue because you actually include the secret `S3_KEY` in your incorrect URL.

timabbott commented:
@whoodes I posted a bunch of comments on the backend, and then made those changes and am merging the backend commit as d4715f23d77b2337168bf1fb2b5a870ef0c1b49d.  There are several high-priority follow-ups that you should do in this PR as they are requirements for us turning on the frontend:

* We need to add `LOCAL_UPLOADS_DIR` support to the `--upload-to-s3` option.  First, a commit should rename it to `--upload` since it's no longer going to be just S3.  And then a commit that changes `export_realm_wrapper` to support upload via the `LOCAL_UPLOADS_DIR` backend (likely, this will be just a `shutil.copyfile` call with the right construction), and then we can remove the AssertionError and fill out the test for the LOCAL_UPLOADS_DIR backend.
* There are several TODOs in `test_realm_export.py` that need to be completed.
* We probably should arrange to store the ID of the `RealmAuditLog` row created in `public_only_realm_export` in the event sent to the deferred_work queue, and then have the queue processor add the `public_url` to the RealmAuditLog event in the `extra_data` field.  This will enable support for deleting the uploaded exports later once they've been accessed.

Thanks @whoodes!

whoodes commented:
@timabbott fyi :)

For 5b87a2e5f711b28fd2f53abe72d31ea6832d4031 I was attempting to test the `extra_data` field's value. I was running into trouble mocking `export_realm_wrapper` in such a way to not run `do_realm_export`, but still receive a return value to be passed into the `extra_data` field.  

I've been reading up on python's unit tests and will continue to look for a solution here.

whoodes commented:
(sidenote: zulipbot isn't updating this PR it seems.)

timabbott commented:
I merged the first commit.  For `export: Add LOCAL_UPLOADS_DIR support to export feature.`, we should be returning the actual URL for the uploaded file -- the whole point of the upload feature is that there's a URL that one can send someone else to import the data.

timabbott commented on zerver/tests/test_realm_export.py:
```diff
@@ -27,69 +21,78 @@ def test_export_as_not_admin(self) -> None:
+        mock_do_write_stats = Mock(spec=do_write_stats_file_for_realm_export,
+                                   return_value=FileNotFoundError)
+        with patch('zerver.lib.export.do_export_realm',
+                   side_effect=mock_do_write_stats()) as mock_export:
```
You can just pass an exception class as `side_effect` without the extra `Mock(...` setup; see for example how we use it in `zerver/tests/test_queue_worker.py`

timabbott commented:
Aside from those two comments, the first few backend commits look correct.  (Though I would have done "Complete TODOs" as 2 separate commits, one for each TODO, since they're fundamentally totally unrelated changes, and it'd be a lot more readable that way)

timabbott commented on zerver/views/public_export.py:
```diff
@@ -25,15 +25,15 @@ def public_only_realm_export(request: HttpRequest, user: UserProfile) -> HttpRes
-
-    RealmAuditLog.objects.create(realm=realm,
-                                 event_type=event_type,
-                                 event_time=event_time)
```
Can you move this reordering to its own commit?  I think it deserves its own commit message.  (I assume you did it to make the testing flow work how you wanted, but I think it is also more correct to do it this way in any case)

whoodes commented:
This is ready for you to take another look at @timabbott :)

I made some additional clean ups to the tests, and split them up into more discrete commits.

(I still need to work on 6628184, but the others commits are good to take a look at.)

timabbott commented on puppet/zulip/files/nginx/zulip-include-frontend/app:
```diff
@@ -83,3 +83,10 @@ location /api/ {
+    include /etc/nginx/zulip-include/api_headers;
+
+    alias /home/zulip/uploads;
+}
```
I don't think it makes sense to add a custom path here.  For one thing, this configuration is shared between the S3 and local file uploads backend.  But also, I think it'd be cleaner to just use the existing directory structure (I don't think `public` in the URL gives the right idea).  I'll post a specific proposal below.

timabbott commented on zerver/lib/export.py:
```diff
@@ -1680,25 +1680,33 @@ def percent_callback(complete: Any, total: Any) -> None:
+        os.makedirs(abs_path)
+        shutil.copy(tarball_path, abs_path)
+        print("Successfully uploaded the tarball to %s" % (abs_path,))
+        public_url = '/public/' + path
```
We should generate URLs that look like just the S3 URLs, with `avatars/` in the URL and everything.  Which means no need for the `/public/` path.

(We will probably do a future refactor to the directories to make the URL nicer, but now isn't the time for that)

timabbott commented on zerver/lib/export.py:
```diff
@@ -1680,25 +1680,33 @@ def percent_callback(complete: Any, total: Any) -> None:
+            bucket=bucket.name,
+            key=key.key)
+        print()
+        print("Uploaded to %s" % (public_url,))
```
So this whole block of code should just be part of the `UploadBackend` classes in `zerver/lib/upload.py`, without the print statements (which should be the same for both code paths).  So this block of code should just be:

```
print("Uploading export tarball")
public_url = upload_backend.upload_export_tarball(...)
print()
print("Uploaded to %s" % (public_url,))
```

with two implementations in the `upload_backend` classes, just like everything else we do for uploading stuff.

timabbott commented:
Posted some comments on the first commit.  Given that everything's stacked on it, I'm just going to fix it and merge.  The `public_url` this output was also wrong -- for the command-line use case, we just want a complete URL you can open in your browser and download.  I then merged the commits through 46790b3280fa8d85dcdcf64e4dea409136380883; see c0ef6c2fc6f34d286b5b7a4fbd811086f98a6392 as I modified that one (and please fix my note about `zerver/lib/upload.py` in follow-ups).  I'll post comments on the rest.

timabbott commented on zerver/worker/queue_processors.py:
```diff
@@ -620,6 +620,11 @@ def consume(self, event: Mapping[str, Any]) -> None:
 
+            # TODO: This enables support for delete after access, and needs to be tested.
+            export_event = RealmAuditLog.objects.get(id=event['id'])
+            export_event.extra_data = public_url
```
Is the idea here that the "delete" feature will reconstruct the actual path from `public_url`?  I guess that's arguably more future-safe for cases where the upload backend gets migrated, so maybe that's the correct approach.  But let's actually write that function in the upload_backend classes and make sure it works (ideally, we'd extend our main end-to-end tests to actually do the fake "upload" of the file, verify it can be downloaded using `self.client_get`, delete it using the function for deleting it, and then verify it isn't still served).  That would give test coverage on the full end-to-end experience, which is definitely missing in this PR (e.g. the `public_url` you returned didn't work if you ran the command locally and then pasted the URL into a browser).

timabbott commented on static/templates/settings/data-exports-admin.handlebars:
```diff
@@ -0,0 +1,18 @@
+        <span class="export_status_text"></span>
+    </div>
+    <form class="form-horizontal">
+        <p class="alert-word-settings-note">{{#tr this}}This tool does a <a href="https://zulipchat.com/help/export-your-organization">public export</a> of your organization's history.{{/tr}}</p>
```
please use `target=_blank` on the /help/ link so it opens in a new tab, and a relative link.  See `static/templates/settings/linkifier-settings-admin.handlebars` for a good example of this sort of note widget.

timabbott commented on static/templates/settings/data-exports-admin.handlebars:
```diff
@@ -0,0 +1,18 @@
+            </div>
+        </div>
+    </form>
+    {{/if}}
```
We should extend this widget in two keys ways:
(1) Add a widget with a list of the exports from RealmAuditLog.
(2) Live-update that list when the export is completed, via the event triggered in `notify_export_completed`

How to do both of those things correctly should be fairly clear from the "New Application Feature" tutorial, plus looking at any of our table-based widgets.  

I think with those completed, we can reasonably make this available to users.

timabbott commented:
@whoodes thanks for your continued work on this.  I merged the first commit.  I think there's two big things that need to happen before this feature is complete:
* The unit test mocking strategy needs to improve.  I think my current thinking is the best option is to move this block into `do_export_realm` (having that function return `tarball_path`):
```
    do_write_stats_file_for_realm_export(output_dir)                                                            
                                                                                                                
    tarball_path = output_dir.rstrip('/') + '.tar.gz'                                                           
    os.chdir(os.path.dirname(output_dir))                                                                       
    subprocess.check_call(["tar", "-czf", tarball_path, os.path.basename(output_dir)])                          
```

If we do that, we can just mock a return value from `do_export_realm` and combine that with the techniques we use in our file-upload unit tests to test the full `export_realm_wrapper` code path.

whoodes commented:
@timabbott I believe this is ready to take a look at :)

whoodes commented:
(Rebased)

timabbott commented:
I merged the first two commits.  I don't understand the testing strategy taken in "test_realm_export: Add coverage for tarball upload and retrieval."; I would have expected us to move to a model with less entangled mocking, not more.  

I think this block:
```
        with patch('zerver.views.public_export.queue_json_publish') as mock_publish:                 
            result = self.client_post('/json/export/realm')                                          
        self.assert_json_success(result)                                                             
        mock_publish.assert_called_once()                                                            
        event = mock_publish.call_args_list[0][0][1]                                                 
        self.assertEqual(mock_publish.call_args_list[0][0][0], 'deferred_work')                      
        self.assertEqual(event['realm_id'], 1)                                                       
        self.assertEqual(event['user_profile_id'], 5)                                                
        self.assertEqual(event['type'], 'realm_exported')                                            
        self.assertTrue(type(event['id']), int)                                                      
                                                                                                     
        with patch('zerver.lib.export.do_export_realm',                                              
                   side_effect=FileNotFoundError) as mock_export:                                    
            with self.assertRaises(FileNotFoundError):                                               
                DeferredWorker().consume(event)                                                      
            args = mock_export.call_args_list[0][1]                                                  
            self.assertEqual(args['realm'], admin.realm)                                             
            self.assertEqual(args['public_only'], True)                                              
            self.assertIn('/tmp/zulip-export-', args['output_dir'])                                  
            self.assertEqual(args['threads'], 6)                                                     

```

should be able to be more like this:
```
diff --git a/zerver/tests/test_realm_export.py b/zerver/tests/test_realm_export.py
index 3d6f06e462..ac1c7ad5f5 100644
--- a/zerver/tests/test_realm_export.py
+++ b/zerver/tests/test_realm_export.py
@@ -23,21 +23,9 @@ class RealmExportTest(ZulipTestCase):
         admin = self.example_user('iago')
         self.login(admin.email)
 
-        with patch('zerver.views.public_export.queue_json_publish') as mock_publish:
+        with patch('zerver.lib.export.do_export_realm') as mock_export:
             result = self.client_post('/json/export/realm')
-        self.assert_json_success(result)
-        mock_publish.assert_called_once()
-        event = mock_publish.call_args_list[0][0][1]
-        self.assertEqual(mock_publish.call_args_list[0][0][0], 'deferred_work')
-        self.assertEqual(event['realm_id'], 1)
-        self.assertEqual(event['user_profile_id'], 5)
-        self.assertEqual(event['type'], 'realm_exported')
-        self.assertTrue(type(event['id']), int)
-
-        with patch('zerver.lib.export.do_export_realm',
-                   side_effect=FileNotFoundError) as mock_export:
-            with self.assertRaises(FileNotFoundError):
-                DeferredWorker().consume(event)
+            self.assert_json_success(result)
             args = mock_export.call_args_list[0][1]
             self.assertEqual(args['realm'], admin.realm)
             self.assertEqual(args['public_only'], True)
```

timabbott commented on zerver/lib/upload.py:
```diff
@@ -573,6 +582,29 @@ def get_emoji_url(self, emoji_file_name: str, realm_id: int) -> str:
+        return public_url
+
+    def delete_export_tarball(self, uri: str) -> bool:
+        path_id = re.sub('https://test-avatar-bucket.s3.amazonaws.com:443/', '', uri)
```
I don't understand this `re.sub` logic; how is it supposed to work?  

I would have expected code involving `urllib.parse` if the plan is to parse a `uri`...

timabbott replied:
(It's also important here that the unit tests can verify the logic for what file we're deleting is correct; the way you have things setup it's a bit unclear to me how this is supposed to work)

whoodes replied:
Basically I was thinking that this just reverse engineers the key from
```
key.key = os.path.join("exports", generate_random_token(32), os.path.basename(tarball_path))
```
When the upload is created in `upload_export_tarball` as this returns a string in the form:
```
exports/4f256889b77c912806eadc1fc28aa813/zulip-export-8vwxgm9r.tar.gz
``` 
 Although, I see how this is a weird and how using `urllib.parse` would be much more appropriate.  But is this sort of approach of grabbing the key from the uri acceptable as a means to delete the file?

I also see how one couldn't verify the correct file is being deleted, I'll rethink this with the tests validating the file.

whoodes replied:
I think the updated approach now addresses this issue:
> (It's also important here that the unit tests can verify the logic for what file we're deleting is correct; the way you have things setup it's a bit unclear to me how this is supposed to work)

(I've also been going off of this comment for reference: https://github.com/zulip/zulip/pull/11996#discussion_r287917477)

timabbott commented on zproject/urls.py:
```diff
@@ -400,6 +400,8 @@
     url(r'^export/realm$', rest_dispatch,
         {'POST': 'zerver.views.public_export.public_only_realm_export'}),
+    url(r'^public_exports$', rest_dispatch,
+        {'GET': 'zerver.views.public_export.get_public_exports'}),
```
This endpoint should be `GET /export/realm`; we'll ultimately want to do `DELETE /export/realm/15` as well for the deletion request.

whoodes replied:
Updated `GET`, holding off on `DELETE` pending further feedback.

whoodes commented:
> I don't understand the testing strategy taken in "test_realm_export: Add coverage for tarball upload and retrieval."; I would have expected us to move to a model with less entangled mocking, not more.

Ahh, right.  I should have realized this when writing it.  I guess I was just thinking about checking the queue parameters.

Would it be more appropriate to still test the queue, just not entangle it within a test that should only be testing the endpoint?

whoodes commented:
Pushed some updates addressing the current batch of comments.

timabbott commented on zerver/lib/export.py:
```diff
@@ -1699,3 +1699,17 @@ def export_realm_wrapper(realm: Realm, output_dir: str,
+    for export in all_exports:
+        exports_dict[export.id] = dict(
+            id=export.id,
+            realm=export.realm.name,
```
I don't think there's a point in having a `realm` value here -- it'll always be for the current realm.

whoodes replied:
Right, updated.

timabbott commented on zerver/lib/export.py:
```diff
@@ -1699,3 +1699,17 @@ def export_realm_wrapper(realm: Realm, output_dir: str,
+            id=export.id,
+            realm=export.realm.name,
+            event_time=export.event_time.ctime(),
+            acting_user=user.full_name,
```
This is wrong -- `user` here is the currently logged-in user; you want `export.acting_user.id` (the frontend should be able to map that to a name).

whoodes replied:
Fixed!

timabbott commented on zerver/tests/test_realm_export.py:
```diff
@@ -95,6 +95,26 @@ def test_realm_export_rate_limited(self) -> None:
+        self.assert_length(export_dict,
+                           RealmAuditLog.objects.filter(
+                               realm=admin.realm,
+                               event_type=RealmAuditLog.REALM_EXPORTED).count())
```
Rather than adding a new test with fake RealmAuditLog data, you should just extend an existing test to do the `client_get` verification; that's more robust to future changes since we're testing the structures we actually create.

whoodes replied:
Cool, I appended this check to both of the endpoint tests.

timabbott commented on zerver/lib/upload.py:
```diff
@@ -788,6 +795,14 @@ def upload_export_tarball(self, realm: Realm, tarball_path: str) -> str:
+    def delete_export_tarball(self, uri: str) -> Optional[str]:
+        # Using `urllib` here would require some extra string manipulation to
+        # remove '/user_avatars/' from the path.
+        path = re.sub('http://zulip.testserver/user_avatars/', '', uri)
```
Reading this, I think we should just store the (relative URL) `path` part, i.e. `/user_avatars/...` in the `extra_data` field for these; this code is going to be broken in the case that the server's URL changes.  And when rendered in a browser as a link, the right thing will happen of automatically inserting the server's URL.

timabbott replied:
(It also lets us get rid of this weird `re.sub` line)

whoodes replied:
Yeah, it's clear as day now how broken this was. Updated.

timabbott commented on zerver/tests/test_upload.py:
```diff
@@ -1757,6 +1772,22 @@ def test_tarball_upload(self) -> None:
+        uri = upload_export_tarball(user_profile.realm, tarball_path)
+        self.assertIsNotNone(uri)
+        self.assertIsNone(delete_export_tarball('not_a_file'))
+        self.assertEqual(delete_export_tarball(uri), urllib.parse.urlparse(uri).path)
```
I'm less sure about these, but I think it'd be better to extend test cases that already "upload" a tarball to do this deletion test, after checking the `get` logic is correct, rather than having an independent test function.  It'll eliminate a lot of semi-duplicated setup code, and again be more robust to changes in how we store things.

whoodes replied:
Agreed, also updated the test name to be clear that we are testing both upload and deletion.

timabbott commented:
I merged a couple commits and posted comments on the next batch.

whoodes commented:
I've addressed the current feedback, @timabbott.

timabbott commented:
This is still a bit of a mess.  I fixed the following issues:
* Changed the names of all the events to be `realm_export`, not `public_export`, since we intend to use this code path for all versions of data export, not just the initial/simplest public-only model.
* The frontend was confusingly and redundantly fetching data on realm exports from two data sources.  I resolved this by removing the data from `page_params`, since this is rarely accessed, potentially sensitive data.  (Also, your implementation gave the full export data to all users, not just admins, so was a security bug).

I think the version I pushed back to this PR is mergable fully working and probably mergable UI-wise assuming we're going to fix-forward quickly from here.  @rishig can you do a quick look at this and see if you agree?  

The following definitely need more work in the UI:
* I marked the fact that the `test_events.py` test doesn't test anything.  To fix it, you need to do some setup that creates a `RealmAuditLog` entry.  Ideally, we'd do something that involves mocking the actual export similar to our other tests, rather than copying RealmAuditLog creation code.  @whoodes can you submit a fix for this?
* We need to add the delete feature.  @whoodes can you submit a fix for this?  We can use the same trashcan styling we use elsewhere for setting this up.
* We should display the timestamps using the same type of nice `timerender` logic we use elsewhere for e.g. "Last active" in the "users" tab.

(And I think we'll have some other UI/styling improvements to make when we add other data export types to this UI).

Finally, there's a few technical issues I'll need to figure out:
* Why we get this exception after triggering the tool:
```
2019-07-26 23:56:36.629 ERR  [django.pylibmc] MemcachedError: error 26 from memcached_get(:1:344e487c3dab5caa0868bc80ad2e3): Transport endpoint is not connected
Traceback (most recent call last):
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django_pylibmc/memcached.py", line 130, in get
    return super(PyLibMCCache, self).get(key, default, version)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/cache/backends/memcached.py", line 79, in get
    val = self._cache.get(key)
pylibmc.Error: error 26 from memcached_get(:1:344e487c3dab5caa0868bc80ad2e3): Transport endpoint is not connected
```
(I have a theory for why this is happening)
* Whether we want to do any limits on how big an organization to export through this flow.

whoodes commented:
> I resolved this by removing the data from page_params, since this is rarely accessed, potentially sensitive data.

Ahh, okay, good to know! This has been a really awesome learning experience, and I'd like to thank you for your patience along the way. 

I'll study the updates you pushed, then begin looking into a fix for the test in `test_events`, as well as adding the delete feature on the frontend.

rishig commented:
Agreed that we can rethink the UI once we have a second export type.

Some quick things in the mean time:
* Let's add a (?) link to /help/export-your-organization to the right of the "Data exports" header.
* Let's call the button "Start public export"
* Below the button let's say "Zulip also supports exporting private streams and messages under some circumstances. [Read more]"
* Remove "Request a public export of your organization's history."
* Ideally clicking "Start public export" would add a row to the table, and in the `File` column we'd just say "Processing .." (rather than having a green success banner). 
Until then, the success message can say "Export started. Check back in a few minutes."

But besides that I agree the UX is mergeable. 

We'll also want to add more text on the page clarifying:
* How often they are allowed to export
* Who (primary admin only?) is allowed to start and view exports
* Some expectation on how long it will take (e.g. "depending on the size of your org, this could take several minutes to an hour")

Thanks @whoodes!

whoodes commented:
Cool, thanks @rishig! I'll draft an update to the UI with those suggestions soon.

(Just as an FYI @timabbott, I'm not completely done with the recent commits I pushed, still in the process of cleaning them up).

timabbott commented:
OK, let me know when this is ready for a review.

whoodes commented:
I believe this is now ready for a round of review.

#### Completed
* Fixed the broken test
* Added the delete feature
  * I think changing the realm export event function to be called `notify_realm_export`
  and then using that for related events is a bit cleaner.  This also makes it clear
  when testing the event, as we can avoid duplicated setup code, while still accurately
  describing the test when covering different events ('noop', 'update', ...). Looking
  forward to feedback regarding this.
  * I also added a dummy file creation function to `test_helpers` as a means to avoid
  further duplication.
* Updated the display of the timestamps using `timerender`
* Added the (?) link
* Changed button to "Start public export"
* Removed "Request a public export..."
* Changed the success banner
* Update the demo GIF

#### Questions
* Not entirely sure where we want [Read more] in "Zulip also supports exporting private streams and messages..." to link to?
* All realm admins are allowed to conduct an export under this model. correct?

#### TODO
* Frontend tests could use improvement
* Not sure if it's a huge deal, but when deleting, with how it's structured right now,
it hits the endpoint in the form:
```
/json/export/realm//user_avatars/exports/1/q9n6V0otNZNKVBtIiaJ4MPi-/zulip-export-5nqldxc5.tar.gz
```
with the double slash after `realm`.

rishig commented:
Great, thanks! For the questions:
* Read more should link to /help/export-your-organization
* Only a primary admin can export. (I forget if the primary admin stuff is merged; if not, then any admin)

timabbott commented:
I merged a few more commits, after making several changes:
* Removed the `op` field; since we're sending all known exports each time, there's no reason to have that.
* Added 4e631876f41397a62e99ad36efa6ff3a4f6acb35 to make that test actually work; your previous code didn't actually run because the user didn't have permission to make the request.

I pushed back to this PR.  Things that still need fixing:
* Rishi's two comments above
* The "No public exports" message when there are done should be "No exports", so we don't need to change that text when we add support for other export types here.
* Can you explain how the "delete" endpoint causes the RealmAuditLog entry to no longer appear?  I don't understand it; I'd have expected the `get` code would need to change to exclude entries with a deletion timestamp.
* We should use `timerender.last_seen_status_from_date` so that it'll give the timing of events that happened today.

whoodes commented:
> Added 4e63187 to make that test actually work; your previous code didn't actually run because the user didn't have permission to make the request.

Ahh okay cool, I tried doing something similar to what you did with `self.do_test` within the `mock.patch` context, but I kept getting `num_events: 0 != 1`, but I never thought to pass `print(self.client_post('/json/export/realm').content)` as the lambda.  I'll do a bit of study there.

> Can you explain how the "delete" endpoint causes the RealmAuditLog entry to no longer appear?

Okay, so what you're saying makes a lot of sense, but I thought under certain circumstances maybe the client would want access to the set of previously deleted exports?  Perhaps this is the wrong thinking.  In any case, this is done when populating the table on the frontend:
```javascript
_.each(exports, function (data) {
        if (data.extra_data.deleted_timestamp === null) {
```
Should I change this to have the backend handle this instead?

whoodes commented:
I addressed Rishi's comments, and updated the demo GIF again.
* I spent the better part of today re-thinking the deletion endpoint.
  * No longer passing the `export_path`
  * The `id` wasn't actually being rendered, fixed that, and changed  the backend to use the `id` instead.
  * Tried to clean things up a bit.
* I now have the export time displaying the times how you suggested.  Took a bit of re-working that--I think--ended up improving things.
* Squashed the frontend deletion code into the main frontend commit.
* Held off an updating the `GET` request to handle filtering the exports.

whoodes commented on zerver/tests/test_events.py:
```diff
@@ -2778,13 +2779,35 @@ def test_notify_realm_export(self) -> None:
         with mock.patch('zerver.lib.export.do_export_realm',
                         return_value=create_dummy_file('test-export.tar.gz')):
             events = self.do_test(
-                lambda: print(self.client_post('/json/export/realm').content),
```
I'm guessing this was just for debugging or something? Looks like I was having trouble by the simple fact of forgetting to assign admin status...

timabbott replied:
Yes, we don't need the print there.

timabbott commented on zerver/lib/export.py:
```diff
@@ -1709,7 +1709,7 @@ def get_realm_exports_serialized(user: UserProfile) -> List[Dict[str, Any]]:
         exports_dict[export.id] = dict(
             id=export.id,
-            event_time=export.event_time.ctime(),
+            export_time=ujson.dumps(export.event_time),
```
I don't understand -- why do we need JSON to send a timestamp?

whoodes replied:
I was looking at this incorrectly.  I realize now that using `dumps` was just a weird way of converting the `datetime` object to a timestamp. Fixed.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5680,3 +5681,16 @@ def notify_realm_export(user_profile: UserProfile) -> None:
+    # However, mypy throws: "AnyStr" of "loads" cannot be "Optional[str]`
+    # if we don't check the None case.
+    export_extra_data = export.extra_data
+    if export_extra_data is not None:
```
I would do `assert export_extra_data is not None` then, not add a useless and confusing if statement.

whoodes replied:
Ahh of course, that should have been obvious. Fixed.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -223,6 +223,7 @@ class OpenAPIArgumentsTest(ZulipTestCase):
         # test because of the variable name mismatch. So really, it's more of a buggy endpoint.
         '/user_groups/{group_id}',  # Equivalent of what's in urls.py
         '/user_groups/{user_group_id}',  # What's in the OpenAPI docs
+        '/export/realm/{export_id}',
```
Move this earlier., it's not related to the `use_groups` block that has comments.

timabbott replied:
In particular, should be next to /export/realm.

whoodes replied:
Woops, silly mistake. Fixed.

timabbott commented on zerver/views/realm_export.py:
```diff
@@ -45,3 +48,17 @@ def export_realm(request: HttpRequest, user: UserProfile) -> HttpResponse:
+                                           realm=user.realm,
+                                           event_type="realm_exported")
+    except RealmAuditLog.DoesNotExist:
+        return json_error(_("The export doesn't exist"))
```
"Invalid data export ID" would be a more standard way to write this.

timabbott commented on zerver/views/realm_export.py:
```diff
@@ -45,3 +48,17 @@ def export_realm(request: HttpRequest, user: UserProfile) -> HttpResponse:
+        return json_error(_("The export doesn't exist"))
+
+    if ujson.loads(export.extra_data).get('deleted_timestamp') is not None:
+        return json_error(_("Export already deleted"))
```
Split the JSON load on a separate line, and do an `if 'deleted_timestamp' in export_data`.

whoodes replied:
Done.

timabbott commented on zerver/views/realm_export.py:
```diff
@@ -45,3 +48,17 @@ def export_realm(request: HttpRequest, user: UserProfile) -> HttpResponse:
+@require_realm_admin
+def delete_realm_export(request: HttpRequest, user: UserProfile, export_id: int) -> HttpResponse:
+    try:
+        export = RealmAuditLog.objects.get(id=export_id,
```
Call this local variable `audit_log_entry`.

whoodes replied:
Done, also updated the tests to use this naming.

whoodes commented:
@timabbott I addressed the batch of comments.  I also split up some clean up commits that should be individually mergeable. Namely:
* Fixing the timestamp thing.
* Changing the `extra_data` key in `export_dict` to have the more consistent, and less ambiguous name `export_data`.
* A general clean up of `test_realm_export`

I also removed `deleted_timestamp` from being added in `queue_processors`. Instead the key, value pair is only added upon deletion.

timabbott commented:
I merged all the backend commits.  I then added 8ac8247dbe2762ef09eeec69187f8619c6261430 and the commit before it to make sure we don't accidentally trigger a DoS (limits are probably conservative) before this has gotten more heavy use.  @rishig you should look at the error message I set for "organization too large"; I think it should be good.

Thanks for making this happen @whoodes!  @rishig we should now be able to update the documentation for data exports to say one can do "public export" self-service.  Not sure if we also need something on /features or what.

I'll open a few technical follow-up issues.

timabbott commented:
https://github.com/zulip/zulip/issues/12992 is the main follow-up issue for this.  We also have some technical infrastructure work to do.

andersk commented:
Also your limit commit failed CI.

whoodes commented:
Added tests for that in #12993
